### ISSUE TITLE
Location: suggested edits

### ISSUE THREAD

**Issue by joshi1983** [hhaccessibility/hhaccessibility.github.io#461]
We want a feature that lets users suggest updates to phone numbers, misspelled location names, addresses and external_web_url's that also manages the risk of someone maliciously changing everything for some sort of April Fool's Day prank.

How to prevent a prankster from doing permanent damage is an open for some discussion.

**Comment by gillbalraj** [comment]
Thanks @joshi1983
 I made 3 suggestions today and received and email from google maps. Stating changes will be made within 48 hours. I'll provide the updates soon.

**Comment by joshi1983** [comment]
@balraj18 did you see any updates?

**Comment by balraj18** [comment]
Well, I haven't checked it so far. I'll update you shortly.

**Comment by gillbalraj** [comment]
I made two changes on residential buildings. These changes are reflecting as expected.

- No verification was required. 

- I received an email yesterday for submitting the details and time it needs to upgrade the changes.

- However, it took a little time to see the changes made on Windsor then on Toronto's building.

![image](https://user-images.githubusercontent.com/26493273/40991022-82e3769c-68c0-11e8-85eb-8502b2803c36.png)

![image](https://user-images.githubusercontent.com/26493273/40990612-68170172-68bf-11e8-8656-b8eb186dbe27.png)
**Last time I made changes on my previous work place(Domino's at Tecumseh Rd), that i doubt if someone updated/changed later on. But i haven't received any notification stating my changes were success/failure**

**Comment by Henry-Hao** [comment]
I added a new "Update Information" button next to the "Accommodation" button which will generate a modal frame asking input from users.
![image1](https://user-images.githubusercontent.com/15081847/42138944-a06e81be-7d53-11e8-8c80-9dd57072ba97.jpg)


![image2](https://user-images.githubusercontent.com/15081847/42138948-add9ce58-7d53-11e8-84a7-2a229610b87c.jpg)


Initially, the input fields will be filled with the existing data from the database. If the user finds that these information is outdated, he can generate a request.
We may need a new table to record all these suggestions, the structure may be something like below:
![image3](https://user-images.githubusercontent.com/15081847/42138978-fcbdf3aa-7d53-11e8-8bf5-810549418e26.jpg)

As for the authenticity of the request, I think there are two ways to verify it.

- Implement another page for the website manager which will list all the suggestions generated from the users and it is the manager who deal with all those.

- Instead of saving the the suggestions in database, they can be sent to the website manager through email.

To deal with pranksters, we can limit the number of suggestions, there are two things we can do from my perspective:
- **Authentication**: We can use a person to verify the information which will decrease the possibility that fake suggestions are accepted.

- **Amount**: We can limit the number of suggestions generated by users to avoid attacks like DoS.

**Comment by joshi1983** [comment]
I like those suggestions, @Henry-Hao.

Can you start by adding the new suggestion table?  To do so, you'll have to add a new migration script file to database/migrations directory and test it using commands like "php artisan migrate:rollback" and "php artisan migrate".

Also, use "id" instead of "suggestion_id" to follow the naming conventions better for the primary key.

**Comment by Henry-Hao** [comment]
@joshi1983 
Sure, i will work on it.
Thank you!

**Comment by joshi1983** [comment]
Henry's database structure changes with the new suggestion table are merged and deployed.  

He's working on the user interface now.

**Comment by joshi1983** [comment]
@Henry-Hao's suggestion feature was merged so suggestions can be submitted and saved.  Some work is remaining to let users review suggestions they submitted and let users with authority to implement the suggestions to review other people's suggestions and potentially implement the changes.

I think for suggestions submitted by the current user or that can be processed by the current user, a link should be added by the red circle here on the profile:
![image](https://user-images.githubusercontent.com/11838191/42739577-a434dfb8-8867-11e8-85be-13e83aa3947a.png)

Henry, what do you think the next screen should look like?  There would be 2 very distinct types of suggestions.  I'm not sure if it should be together in a single list or broken into 2 separate pages.

**Comment by Henry-Hao** [comment]
Thank you, Josh.

Since the other two tabs 'My Reviews' and 'My Locations' are designed to open a new page i would suggest  apply the same strategy to Suggestions.

For all of the three tabs, maybe the example below would gives a better choice.
<https://www.w3schools.com/bootstrap/tryit.asp?filename=trybs_tabs_dynamic&stacked=h>

**Comment by Henry-Hao** [comment]
For the part of 'Suggestions of My Locations' I think we can simply put a link to a column in the My Locations list which would redirect the page to all the suggestions related to this specific location add by the current user. And for the 'My Suggestions' we can list them the same way as 'My Locations' does.

Also I totally agree with you about storing the information between the suggestions and the reviewers. The data may be useful in the futuer if we want to limit user's permission based on the accuracy of location he added.

For the UI part, i think we can have a list of the suggestion and after user clicks it. It will expand and show the original information on the left while the suggestions on the right. For each line we can add a new button to the end of the line and the bottom of the ```<div>``` which enables user to accept the suggestion for a specific line or the whole suggestion. Just as you said above.

**Comment by joshi1983** [comment]
A separate page for reviewing a specific suggestion would be better for screen readers and should keep the features separated better.  The modal is good for most users but using a modal requires us to eventually make a full page alternative anyway for people using screen readers.

**Comment by Henry-Hao** [comment]
@joshi1983 
I created a mock up image for the suggestion review.
Could you please give me some advice, thank you?
![ac](https://user-images.githubusercontent.com/15081847/43019606-8f368392-8c2b-11e8-874b-3be19b6ad5a6.jpg)

**Comment by Henry-Hao** [comment]
@joshi1983 
I'm sorry but actually I didn't get it. Please correct me if I'm wrong. 
There are two pages for different functionalities, one for the user to review all the suggestions binded to the locations which are added by the current user, which means the creator (the current user) has the authority to accept or drop all these suggestions related to this location. In this case, the user can access the page above by clicking the column in "My Locations" which indicating the number of suggestions of each location.
Another one is for the user to review all the suggestions made by himself and check the status of these suggestions, whether they are accepted or denied. In this case there will be another link next to "My Locations" and "My Reviews" in Profile page.

**Comment by joshi1983** [comment]
Does this clarify how it would work?  We could have a call about it.

My Locations: would have a column saying how many suggestions are pending for each location.  The number of suggestions can link to "List Page".  The link would specify the location id so the other page can list suggestions specifically for 1 location.

List Page: lists all suggestions for a single location.  Each row links to Suggestion Review.

Suggestion Review: a page where you review 1 suggestion.  You would accept any changes you agree with or mark them as reviewed or deleted.

**Comment by Henry-Hao** [comment]
Yes, I got it. So basiclly I don't need to list the detailed information for each suggestion in the List Page. Instead, a link of the item on List Page will redirect to a page which shows it.
Thank you, this helps a lot.

**Comment by joshi1983** [comment]
The new location edit feature was merged with pull request #661.  Get latest with your work since it could conflict with your changes.

When a user is looking at a location he created, the "Suggest Updates" button becomes "Update" and links to the new location edit feature instead of the suggestion dialog.  Other than that, a few routes changed in location.php which might affect how you wrote some new suggestion-related routes.

**Comment by joshi1983** [comment]
The latest changes from Henry have been merged and deployed to https://app.accesslocator.com.  

There are some improvements that #669 is covering.

@singlaneha2205 can you try the new suggestion feature out and see if you can think of improvements beyond what #669 covers?

**Comment by joshi1983** [comment]
I unassigned @singlaneha2205 to encourage any volunteers to test this.
